model/device.go

```diff
@@ -78,6 +133,100 @@ func (da DeviceAttribute) Validate() error {
 	)
 }
 
+func allowedType(v reflect.Value) bool {
+	if v.Type().Kind() == reflect.Int {
+		return true
+	}
+	if v.Type().Kind() == reflect.String {
+		return true
+	}
+	if v.Type().Kind() == reflect.Float64 {
+		return true
+	}
+	return false
+}
+
+func reflectValuesEqual(rVal1 reflect.Value, rVal2 reflect.Value) bool {
+	if rVal1.Len() != rVal2.Len() {
+		return false
+	}
+	for i := 0; i < rVal1.Len(); i++ {
+		if rVal1.Index(i).Kind() != rVal2.Index(i).Kind() {
+			return false
+		}
+		if !allowedType(rVal1.Index(i)) || !allowedType(rVal2.Index(i)) {
+			if !reflect.DeepEqual(rVal1.Index(i), rVal2.Index(i)) {
+				return false
+			}
+		} else {
+			if rVal2.Index(i).Kind() == reflect.Int {
+				value1 := rVal1.Index(i).Interface().(int)
+				value2 := rVal2.Index(i).Interface().(int)
+				if value1 != value2 {
+					return false
+				}
+			}
+			if rVal2.Index(i).Kind() == reflect.String {
+				value1 := rVal1.Index(i).Interface().(string)
+				value2 := rVal2.Index(i).Interface().(string)
+				if value1 != value2 {
+					return false
+				}
+			}
+			if rVal2.Index(i).Kind() == reflect.Float64 {
+				floatComparePrecision := "%.8f"
+				floatValue1 := rVal1.Index(i).Interface().(float64)
+				floatValue2 := rVal2.Index(i).Interface().(float64)
+				value1 := fmt.Sprintf(floatComparePrecision, floatValue1)
+				value2 := fmt.Sprintf(floatComparePrecision, floatValue2)
+				if value1 != value2 {
+					return false
+				}
+			}
+		}
+	}
+	return true
+}
+
+func (da DeviceAttribute) Equal(e DeviceAttribute) bool {
+	if (da.Value == nil || e.Value == nil) && !(da.Value == nil && e.Value == nil) {
+		return false
+	}
+	rVal1 := reflect.ValueOf(da.Value)
+	rVal2 := reflect.ValueOf(e.Value)
+	if rVal1.Kind() != rVal2.Kind() {
+		if rVal1.Kind() == reflect.Slice && rVal2.Kind() == reflect.Array ||
+			rVal1.Kind() == reflect.Array && rVal2.Kind() == reflect.Slice {
```

Comment:
`json.Unmarshal` always unmarshal JSON arrays to slices (`[]interface{}`). The check for `reflect.Array` is redundant.

Reply:
yes I know, but in our tests we really have slices and arrays explicite, I made this condition really to pass all the test cases we already have.